### ISSUE TITLE
Bug: Candid encoding not correctly assigning different variants for a `key: variant val` vector

### ISSUE THREAD

**Issue by ozwaldorf** [dfinity/sdk#2685]

**Comment by chenyan-dfinity** [comment]
I think the problem is that `dfx` is not using the did file when you make canister calls. This can happen if you call the canister by its canister id, not the name defined in `dfx.json`, or you are calling an external canister.

Can you try `dfx canister call <canister_id> <method> --candid <.did file>`, and see if it works?

As for the real fix, I think `dfx` should always try to fetch the did file from metadata before sending message.

**Comment by ozwaldorf** [comment]
> Can you try `dfx canister call <canister_id> <method> --candid <.did file>`, and see if it works?

Manually linking the candid is a valid fix

> I think the problem is that `dfx` is not using the did file when you make canister calls. This can happen if you call the canister by its canister id, not the name defined in `dfx.json`, or you are calling an external canister.

So it's actually weird, for deserialization of the response, dfx will correctly use the candid even if it's referenced by the id directly. It seems like for serialization, it's not linking the candid like deserialize does


> As for the real fix, I think `dfx` should always try to fetch the did file from metadata before sending message.

This is a good idea. But still it's efficient to use the local candid to serialize like mentioned above

**Comment by sesi200** [comment]
Ok, let's call it a bug for now, and after the .did fetch is fixed convert it into a feature request for the metadata fetch. Internal ticket: [SDK-783](https://dfinity.atlassian.net/browse/SDK-783)

**Comment by chenyan-dfinity** [comment]
> So it's actually weird, for deserialization of the response, dfx will correctly use the candid even if it's referenced by the id directly. It seems like for serialization, it's not linking the candid like deserialize does

No. Currently, when calling canister by id, did file is not used. You can confirm this by observing that the decoded record field is always a hash id instead of a string name. The difference is that the current serialization implementation without type is incomplete, it can produce a wrong type and causing errors with deserialization in the future. But we never encourage to use serialization without types, so we have no plans to fix this corner case.

With deserialization, you can always decode the message properly, but with types, you get a nicer field names.

**Comment by ozwaldorf** [comment]
> > So it's actually weird, for deserialization of the response, dfx will correctly use the candid even if it's referenced by the id directly. It seems like for serialization, it's not linking the candid like deserialize does
> 
> No. Currently, when calling canister by id, did file is not used. You can confirm this by observing that the decoded record field is always a hash id instead of a string name. The difference is that the current serialization implementation without type is incomplete, it can produce a wrong type and causing errors with deserialization in the future. But we never encourage to use serialization without types, so we have no plans to fix this corner case.
> 
> With deserialization, you can always decode the message properly, but with types, you get a nicer field name.

![image](https://user-images.githubusercontent.com/8976745/196239000-bc146bd5-9525-49d4-81d0-a50224ae7c08.png)
Seems like it does a reverse lookup to the project for the types on responses?
